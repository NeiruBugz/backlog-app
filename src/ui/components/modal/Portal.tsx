import { ReactNode } from 'react';
import ReactDOM from 'react-dom';

const Portal = ({ children }: { children: ReactNode }) => {
  const createOverlay = (): Element => {
    const overlay = document.createElement('div');
    overlay.setAttribute('id', 'overlay');
    document.body.appendChild(overlay);
    return overlay;
  };

  const overlay: Element = document.getElementById('overlay')
    ? document.getElementById('overlay') as Element
    : createOverlay();

  return ReactDOM.createPortal(children, overlay);
};

export { Portal };