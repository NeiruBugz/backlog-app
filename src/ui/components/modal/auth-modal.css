.ba-auth-form {
  display: flex;
  flex-direction: column;
  font-size: 18px;
}

.ba-auth-form-label {
  margin: 8px 0;
}

.ba-auth-form-input {
  margin-top: 8px;
  border-top: 0;
  border-left: 0;
  border-right: 0;
  outline: 0;
}

.ba-auth-form-buttons {
  margin-top: 16px;
  display: flex;
  justify-content: space-between;
}
