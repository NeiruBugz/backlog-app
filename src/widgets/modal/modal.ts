import { atom } from 'nanostores';

interface ModalStore {
  id: string | null;
  isVisible: boolean;
  modalProps?: unknown;
}

const initialValue: ModalStore = {
  id: '',
  isVisible: false,
};

const modal = atom(initialValue);

const setModal = function setModalId(payload: ModalStore) {
  modal.set(payload);
};

export { modal, setModal };
