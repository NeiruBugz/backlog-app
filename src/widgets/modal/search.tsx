import { useCallback, useEffect, useState, forwardRef } from 'react';
import { useTranslation } from 'react-i18next';

import type { ChangeEventHandler } from 'react';
import type { HowLongToBeatEntry } from 'howlongtobeat';
import { api } from '@shared';
import { SearchResultsList } from '@widgets';

const SearchModal = forwardRef<HTMLDivElement, unknown>((_, ref) => {
  const { t } = useTranslation();
  const [query, setQuery] = useState('');
  const [searchResults, setSearchResults] = useState<HowLongToBeatEntry[]>([]);

  const onGameSearch = useCallback((value: string) => {
    if (!value.length) {
      setSearchResults([]);
      return;
    }

    api
      .search(value)
      .then((result) => {
        setSearchResults(result);
      })
      .catch((err) => {
        console.error(err);
      });
  }, []);

  useEffect(() => {
    const onEnterPress = (e: KeyboardEvent) => {
      if (e.key === 'Enter' && query !== '') {
        onGameSearch(query);
      }
    };

    document.addEventListener('keypress', onEnterPress);

    return () => {
      document.removeEventListener('keypress', onEnterPress);
    };
  }, [onGameSearch, query]);

  const onInputChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    setQuery(event.currentTarget.value);
  };

  const onButtonPress = () => {
    onGameSearch(query);
  };

  return (
    <div className="w-1/2 h-1/2" ref={ref}>
      <div className='flex'>
        <input
          type="search"
          className="input input-bordered input-primary w-full mr-2"
          placeholder={t('home.search.inputPlaceholder') || ''}
          value={query}
          onChange={onInputChange}
        />
        <button className="btn btn-primary" onClick={onButtonPress} disabled={query.length === 0}>
          {t('home.search.button')}
        </button>
      </div>
      <SearchResultsList results={searchResults} />
    </div>
  );
});

SearchModal.displayName = 'SearchModal';

export { SearchModal };
