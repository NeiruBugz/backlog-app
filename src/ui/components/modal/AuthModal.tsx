import { Modal } from 'antd';
import './auth-modal.css';
import { login } from '@store';
import { Button, LabeledInput } from '@components';

const AuthModal = ({ isOpen, onCancel }: { isOpen: boolean, onCancel: () => void; }): JSX.Element | null => {
  return isOpen ? (
    <div className="ba-modal-overlay">
      <Modal centered footer={null} title='Sign In' className="ba-modal" open={isOpen} onCancel={onCancel}>
        <div className="ba-modal">
          <header className="ba-modal-header">Sign In</header>
          <div className="ba-modal-body">
            <form action="submit" className="ba-auth-form">
              <LabeledInput
                label='Login'
                inputType="text"
                inputId="login"
                inputName="login"
                labelClass='ba-auth-form-label'
                inputClass='ba-auth-form-input'
              />
              <LabeledInput
                label='Password'
                inputType="password"
                inputId="password"
                inputName="password"
                labelClass='ba-auth-form-label'
                inputClass='ba-auth-form-input'
              />
              <div className="ba-auth-form-buttons">
                <Button mode="primary" onClick={login} text='Sign in' />
                <button className="ba-button ba-button--secondary">Cancel</button>
              </div>
            </form>
          </div>
        </div>
      </Modal>
    </div>
  ) : null;
};

export { AuthModal };
