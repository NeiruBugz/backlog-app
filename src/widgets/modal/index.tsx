import { useRef, useEffect, useCallback } from 'react';
import { useStore } from '@nanostores/react';
import ReactDOM from 'react-dom';

import { modal, setModal } from './modal';
import { SearchModal } from './search';

import type { ReactPortal } from 'react';
import { useOutsideClick } from '@shared';

const Modal = (): ReactPortal | null => {
  const { isVisible, id } = useStore(modal);
  const modalBodyRef = useRef<HTMLDivElement>(null);

  const onModalClose = useCallback(() => setModal({ id: null, isVisible: false }), []);

  useEffect(() => {
    const onEscapePress = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        onModalClose();
      }
    };

    document.addEventListener('keydown', onEscapePress);

    return () => {
      document.removeEventListener('keydown', onEscapePress);
    };
  }, [onModalClose]);

  useOutsideClick(modalBodyRef, onModalClose);

  const modalComponent = () => {
    if (id === 'search') {
      return <SearchModal ref={modalBodyRef} />;
    }

    return <div ref={modalBodyRef}>Modal</div>;
  };

  if (!isVisible) {
    return null;
  }

  return ReactDOM.createPortal(
    <div id="modalOverlay">
      <div
        className="w-[100vw] h-[100vh] z-50 bg-slate-300/75 absolute top-0 left-0 flex items-center justify-center"
        ref={modalBodyRef}
      >
        {modalComponent()}
      </div>
    </div>,
    document.body
  );
};

export { Modal };
