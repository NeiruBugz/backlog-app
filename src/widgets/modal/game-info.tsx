import type { HowLongToBeatEntry } from 'howlongtobeat';
import { forwardRef, useState } from 'react';
import { GameItem, nanoGames } from '@entities';
import { Divider } from 'widgets/divider';

import type { ChangeEventHandler, MouseEventHandler } from 'react';
import { updateGameDocument } from '@shared';
import { useStore } from '@nanostores/react';

const GameInfo = forwardRef<HTMLDivElement, HowLongToBeatEntry>((props, ref) => {
  const game = useStore(nanoGames).find((game) => game.title === props.name);
  const [review, setReview] = useState('');
  const [isEditMode, setEditMode] = useState(false);

  const onReviewChange: ChangeEventHandler<HTMLTextAreaElement> = ({
    currentTarget: { value },
  }) => {
    setReview(value);
  };

  const onReviewSubmit: MouseEventHandler<HTMLButtonElement> = () => {
    if (game) {
      updateGameDocument({ id: game.id, field: { key: 'review', value: review } }, 'games').then(
        () => setEditMode(false)
      );
    }
  };

  const enableEditMode: MouseEventHandler<HTMLButtonElement> = () => {
    if (game && game.review) {
      setReview(game?.review);
      setEditMode(true);
    }
  };

  const disableEditMode: MouseEventHandler<HTMLButtonElement> = () => setEditMode(false);

  return (
    <div className="h-10/12 max-w-10/12 rounded-box min-w-fit bg-white p-4" ref={ref}>
      <GameItem item={props} />
      {game?.review && !isEditMode ? (
        <>
          <Divider>Review</Divider>
          <p className="prose p-4">{game?.review}</p>
          <button className="btn-ghost btn" onClick={enableEditMode}>
            Edit
          </button>
        </>
      ) : (
        <>
          <Divider>Write a Review</Divider>
          <textarea
            value={review}
            onChange={onReviewChange}
            className="textarea-bordered textarea w-full text-[1.2rem]"
          />
          <button className="btn-primary btn" onClick={onReviewSubmit}>
            Submit
          </button>
          {isEditMode ? (
            <button className="btn-warning btn ml-2" onClick={disableEditMode}>
              Cancel
            </button>
          ) : null}
        </>
      )}
    </div>
  );
});

GameInfo.displayName = 'GameInfo';

export { GameInfo };
