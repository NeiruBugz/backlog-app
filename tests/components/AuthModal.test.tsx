import React from 'react';
import { render } from '@testing-library/react';
import { AuthModal } from '../../src/ui/components/modal/AuthModal';

describe('AuthModal Component', () => {
  test('snapshot', () => {
    render(<AuthModal isOpen={true} />);
    expect(screen).toMatchSnapshot();
  });
});