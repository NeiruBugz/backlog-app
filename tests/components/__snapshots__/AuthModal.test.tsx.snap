// Vitest Snapshot v1

exports[`AuthModal Component > snapshot 1`] = `
<div>
  <div
    class="ba-modal-overlay"
  />
</div>
`;
